[quant][core][gpu][bug fix] Fixed off by one index issue in broadcasted_bias

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75418
* #75447
* __->__ #75483
* #75446

Summary:
There was an off by one index issue in new_size (used for broadcasted_bias).
This has now been corrected.

The matrix multiplication's output's last dimension is the number of out features, which is the
size of bias. We create `new_size` for `broadcasted_bias`, whom we want to have the same number
of dimensions as the matmul output and for it have the same size for the last dimension for broadcasting purposes.
Previously, we had `new_size[1] = bias_.value().size(0);`, but this is wrong, in general.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35489965](https://our.internmc.facebook.com/intern/diff/D35489965)

dzdang committed:
[quant][core][gpu][bug fix] Fixed off by one index issue in broadcasted_bias

Summary:
There was an off by one index issue in new_size (used for broadcasted_bias).
This has now been corrected.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35489965).

dzdang committed:
Update on "[quant][core][gpu][bug fix] Fixed off by one index issue in broadcasted_bias"


Summary:
There was an off by one index issue in new_size (used for broadcasted_bias).
This has now been corrected.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35489965](https://our.internmc.facebook.com/intern/diff/D35489965)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35489965).

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35489965).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -69,11 +69,11 @@ void PackedLinearWeightCudnn::apply_impl_helper(const at::Tensor& quantized_outp
     // the number of trailling dimensions is quantized_output.dim() - 2. We also prepend a leading dimension for clarity
     std::vector<int64_t> new_size(quantized_output.dim(), 1);
-    new_size[1] = bias_.value().size(0);
+    new_size.back() = bias_.value().size(0);
```
do we need to change the size of `new_size` now? can it be the same size as bias?

dzdang replied:
I need to check if broadcast addition works in cudnn 8.4.0, if it does, then we don't need new_size

dzdang replied:
@jerryzh168  Channel broadcasting addition seems to work, but both conv and linear are running into that `uid` issue I mentioned on slack. I think maybe we should keep the old format (keep `new_size` for now) until NVIDIA figures out what the issue is? Or I could implement it acknowledging the `uid` issue (it's not blocking, but we will need frontend `v0.6` for it because the buffer size for the previous frontend is fixed at 10. we need 11 right now due to the problem I identified)

jerryzh168 replied:
I see, can we try upgrading to `v0.6`? I saw it is reverted, what is the plan now?

dzdang replied:
@jerryzh168 It looks like cudnn updated the frontend to address the return statement error: https://github.com/NVIDIA/cudnn-frontend/commit/fa611998a360cbabaa2dcc7c9859748144114fc0

I'll try re-landing the PR

jerryzh168 replied:
btw why do we change from new_sizes[1] to new_sizes.back() here? could you explain in PR summary?

dzdang committed:
Update on "[quant][core][gpu][bug fix] Fixed off by one index issue in broadcasted_bias"


Summary:
There was an off by one index issue in new_size (used for broadcasted_bias).
This has now been corrected.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35489965](https://our.internmc.facebook.com/intern/diff/D35489965)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][bug fix] Fixed off by one index issue in broadcasted_bias"


Summary:
There was an off by one index issue in new_size (used for broadcasted_bias).
This has now been corrected.

The matrix multiplication's output's last dimension is the number of out features, which is the
size of bias. We create `new_size` for `broadcasted_bias`, whom we want to have the same number
of dimensions as the matmul output and for it have the same size for the last dimension for broadcasting purposes.
Previously, we had `new_size[1] = bias_.value().size(0);`, but this is wrong, in general.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35489965](https://our.internmc.facebook.com/intern/diff/D35489965)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35489965).

jerryzh168 reviewed (approved):
I see, I thought it was off by one issue in cudnn implementation before

dzdang committed:
Update on "[quant][core][gpu][bug fix] Fixed off by one index issue in broadcasted_bias"


Summary:
There was an off by one index issue in new_size (used for broadcasted_bias).
This has now been corrected.

The matrix multiplication's output's last dimension is the number of out features, which is the
size of bias. We create `new_size` for `broadcasted_bias`, whom we want to have the same number
of dimensions as the matmul output and for it have the same size for the last dimension for broadcasting purposes.
Previously, we had `new_size[1] = bias_.value().size(0);`, but this is wrong, in general.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35489965](https://our.internmc.facebook.com/intern/diff/D35489965)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35489965).
